design-proposals/2022-11-29-metal-port.md

```diff
@@ -0,0 +1,447 @@
+
+The goal here is to move as many setup stages as possible off the main thread.  That should include all the buffer allocation and setup and some of the indirect command buffer encoding.  Direct encoding is going to still occur on the main thread.
+
+#### Required Changes
+
+It's not clear what changes will need to be made here.  The process goes like this:
+- Find a good test case.  Usually something that involves a lot of zooming and panning.
+- Run the test case while recording one or more of: memory or CPU
+- Look for a bottleneck
+- Fix the bottleneck
+
+Typically bottlenecks are where the main CPU is waiting on something it shouldn't be, or tiles are only being loaded on a single thread or other similar issues.  It's never obvious until you start staring at the traces.
+
+#### Benefits
+
+MapLibre Native makes some use of threads, but not as much as it should.  With OpenGL it's not using shared contexts, though it does a lot of work in the Layers to prep for the GL work that has to be done on the main thread.  So it's not terrible, but it obviously could be better.
+
+Mobile devices, even the really cheap ones, now have a lot of CPU cores and a lot of opportunity for work to be done off the main thread.  Maps have an embarassing amount of high level parallelism.  It's worth periodically analyzing the performance to see what might be updated for data loading.
+
+The benefits are faster loading, lower latency, better responsiveness to the user.
+
+### Shader optimization
+
+The first version of shaders will work with the direct rendering in Metal.  They'll be simple manual translations from the GLSL in the existing toolkit and they'll look and act similarly to the old shaders.
+
+The second version of the shaders will consist of changes to support indirect rendering.  The way buffers are passed in is different when doing indirect rendering in a way that's not obvious.
+
+Further, this is the time to break the shaders to modular pieces to be reused by developers outside the toolkit.  Passing in a new shader to replace existing functionality works a lot better when you can call sub-routines to implement that MapLibre Native basics for you.
+
+#### Required Changes
+
+This pass is the last version of the shaders and would consist of:
+- Modularization.  We can make it easier to reuse shaders by sharing code where useful.  Even putting the core of a given shader into a subroutine would make it easier to reuse.
+- Data driven styling.  How much of that logic can be pushed into the shaders?  The more work they do, the less the CPU does.
+- Optimization.  Apple provides tools for analyzing time spent in the shaders.  Some optimization here is always useful.
+- Packaging.  Normally Metal shaders are pre-compiled and shipped with the app binary.  We may want to support a pure source version.
```

Comment:
Personally, I've moved away from generating shaders on the fly.  It never worked out as well as I liked, even in OpenGL, and it was incredibly confusing to debug.

These days I just use the standard flow control in the shader itself.  It makes things so much easier to develop and maintain and doesn't seem to have significant performance implications in these sorts of use cases.  Getting memory in and out tends to dominate.  The cost of my shader calculations largely don't.

I suspect we can do data driven styling in Metal without resorting to shader generation.  It's an open question but let's find out!

Reply:
It will be interesting to find out!

Reply:
> It would be more interesting to know wether the IR can be modified / how we can generate shaders on the fly.
> This will be required for data driven styling with good performance.

I wouldn't dare mess with precompiled metal shaders, lest Apple decide to randomly change the implementation without telling anyone. And why would they?

> You will likely still need shader generation for OpenGL because of attribute limits (we have customers with only 15 attributes), so the renderer modularization probably has to respect that.

In designing a library like this, one would do best to constrain themselves to the minimum guaranteed device capabilities defined by your RHI. Assuming the worst from the start is preferable. I have to wonder what you would be doing that needs that many attributes, the fixed function stage between your vertex and fragment shader is going to be doing a lot of interpolation. Usually that is an indicator that what you're doing might not be the most efficient way to go about it.